### ISSUE TITLE
False negative when trying to access a field of TypedDict defined with `total=false`

### ISSUE THREAD

**Issue by yurijmikhalevich** [microsoft/pyright#1584]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
pyright doesn't error when accessing optional properties of TypedDict defined with `total=False`

**To Reproduce**
```python
from typing import TypedDict


class RequiredMyClass(TypedDict):
    fieldA: str


class MyClass(RequiredMyClass, total=False):
    fieldB: int

# pyright correctly understand that fieldA is required in MyClass, when fieldB is not

instance: MyClass = {'fieldA': 'test'}
print(instance['fieldB'])   # no error here
```

**Expected behavior**
Error when trying to access an optional TypedDict field defined with `total=False`

**VS Code extension or command-line**
Both, with the same result. pyright in command line v1.1.117 and v1.1.118.

**Comment by erictraut** [comment]
Pyright is doing the right thing according to [PEP 589](https://www.python.org/dev/peps/pep-0589/). When a TypedDict is marked "total=False", it indicates that the fields within that class are not required. If they are present, it defines their type.

In your example above, you are trying to access `fieldB` which is a defined field, and its type (if present) is `int`. A type checker should assume that it is potentially present in your dict and not generate an error if you attempt to access it. In contrast, if you attempt to access a field `fieldC`, pyright will generate an error.

I confirmed that pyright's behavior is consistent with other type checkers like mypy.

**Comment by yurijmikhalevich** [comment]
@erictraut Thank you for the response.

Is there a way to type a field as ommitabble in a TypedDict and have type guards?

**Comment by erictraut** [comment]
No, that's not part of the TypedDict design.

One solution is to make all fields required but allow them to contain `Optional` types — i.e. allow then to contain a `None` value as a "not present" indicator. Type checkers (including pyright) will verify that you check for `None` before using the value inappropriately.

**Comment by yurijmikhalevich** [comment]
@erictraut from the PEP 589:

```
Type checkers **may** allow reading an item using d['x'] even if the key 'x' is not required, instead of requiring the use of d.get('x') or an explicit 'x' in d check. The rationale is that **tracking the existence of keys is difficult to implement in full generality**, and that disallowing this could require many changes to existing code.

The exact type checking rules are up to each type checker to decide. In some cases potentially unsafe operations may be accepted if the alternative is to generate false positive errors for idiomatic code.
```

PEP 589 accepts the current pyright behavior but indicates that it's not optimal and only accepted due to the performance concerns. I belive that forcing user to access `total=False` fields via `.get('x')` as suggested in the docs will be a nice optional rule to have.

**Comment by yurijmikhalevich** [comment]
@erictraut 

>  allow then to contain a None value as a "not present" indicator.

It's not the same. `fieldA: Union[T, None]` is not the same as ommitable `fieldA: T`, in the first case access `dict['fieldA']` is totally safe, in the second case it's unsafe.

**Comment by erictraut** [comment]
Ah, I didn't remember reading that in the PEP. It's not clear to me, however, which conditional expression would be used as a guard in this case. You can't use `if instance['fieldB']:` because that will generate an exception if `fieldB` is omitted. The alternatives are to 1) enclose the check in a `try` block, which is no longer a simple type guard construct, or 2) use something like instance.get('fieldA') and check it against None.

Yes, I understand that omittable is not the same as required but possibly None. I'm just saying that if you craft your TypedDicts as having optionals rather than omittables, your code will be safer and easier to maintain.

**Comment by yurijmikhalevich** [comment]
@erictraut do you think that rule forcing user to access `total=False` fields via `.get('x')` as suggested in the docs will be a nice optional rule to have?

And then you can write `if instance.get('fieldB')`.

**Comment by erictraut** [comment]
Let me check with the authors of PEP 589. I think they also implemented the support in mypy, so I'm curious why they opted not to implement this restriction.

**Comment by yurijmikhalevich** [comment]
@erictraut sounds great! I'm curious about that too. Can you let me know their response? Thank you!

**Comment by yurijmikhalevich** [comment]
@erictraut here is what I found in the [mypy docs](https://mypy.readthedocs.io/en/stable/more_types.html#totality): `Requiring get() everywhere would be too cumbersome.` But it doesn't say much.

**Comment by erictraut** [comment]
Here's my post in the typing-sig: https://mail.python.org/archives/list/typing-sig@python.org/thread/CCEFKMRGLCBCZROMLTTGETRQKVQRZWDP/

**Comment by yurijmikhalevich** [comment]
@erictraut thank you.

**Comment by layday** [comment]
> however, which conditional expression would be used as a guard in this case

The "idiomatic" guard expression here would be `"fieldB" in instance`.

**Comment by yurijmikhalevich** [comment]
@erictraut mypy dev responded that they are going to add a strictness flag to enforce the rule https://mail.python.org/archives/list/typing-sig@python.org/message/6Z2QLMHYHUJOEZRURIGTAGDJ4KZERPUQ/.

> various variants of this where a technically non-required key is always present. Also users tend to just use total=False even if some keys are always present

These rationalizations sound to me like a user's problem which will occur if they incorrectly type their code.

**Comment by erictraut** [comment]
Yeah, I agree. TypedDict is still not in wide use (it was standardized only recently), so I don't think this would break much existing code if we were to enforce this rule. So I'm inclined to add a diagnostic check for this case. However, there are still a couple of open questions that need to be considered:

1. Should this check be added as a new, separate rule (e.g. "reportNonrequiredTypedDictAccess")? Or should it be part of the more general "reportGeneralTypeIssues" rule? There's a downside to adding more configuration knobs, especially those that are extremely specific, so I tend to prefer using existing diagnostic rule categories unless there's a really compelling need to control these diagnostic checks independently. I guess I could start by adding it to the "reportGeneralTypeIssues" category and pull it out later if there's sufficient demand.
2. Is there any form of type narrowing that needs to be added in conjunction with this change? As @layday mentioned above, the guard expression `"fieldB" in instance` is the obvious choice. But we have a problem here because this doesn't fit the normal pattern of type narrowing, since the type of `instance` remains the same before and after the conditional check. Adding this support would require a different type of infrastructure than the current type narrowing mechanism in pyright, and that has consequences in terms of complexity, performance, and maintainability. So I'd prefer to avoid adding it unless/until we know that there's sufficient demand. So the question is whether adding the diagnostic check without adding a "narrowing-like" mechanism is OK?

What do you think?

**Comment by erictraut** [comment]
I'll also point out that [PEP 655](https://www.python.org/dev/peps/pep-0655/) (which I'm hoping will be ratified in time for Python 3.10) should eliminate the concern about "users tend to just use total-False". Pyright already has support for PEP 655.

**Comment by erictraut** [comment]
Good news. I figured out how to add support for the "S in D" narrowing pattern such that accesses to non-required fields are allowed when used within that conditional block. So that removes question 2 above. As for question 1, I've chosen to include it in the "reportGeneralTypeIssues" category for now.

This support will be in the next release of pyright.

**Comment by yurijmikhalevich** [comment]
@erictraut

> I guess I could start by adding it to the "reportGeneralTypeIssues" category and pull it out later if there's sufficient demand.

I agree with your point and the decision. I suggested having an independent rule initially only to make the introduction of the new rule softer.

> the guard expression "fieldB" in instance is the obvious choice

`"fieldB" in instance` is indeed idiomatic. I'm glad that you've found a way to add support for it in pyright.

My suggestion is
1. To support `if "fieldB" in instance` that will narrow the type and allow for `instance["fieldB"]` access.
2. To allow accessing the field via `instance.get("fieldB")`. In this case the type of returned value should be `Optional[Type]` to indicate the possibility of it being `None`.

> I'll also point out that PEP 655 (which I'm hoping will be ratified in time for Python 3.10) should eliminate the concern about "users tend to just use total=False".

These were my thoughts exactly. I'm excited about the PEP 655!

> This support will be in the next release of pyright.

Thank you! I look forward to it!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.120, which I just published. It will also be included in the next release of pylance.

**Comment by yurijmikhalevich** [comment]
@erictraut I saw the changelog! Happy to see it shipped! Thanks!
